edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/Handlers/ConnectionHandler.cs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Threading.Channels;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Device;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ConnectionHandler : IConnectionRegistry, ISubscriber, IMessageConsumer
+    {
+        const string TopicDeviceConnected = "$edgehub/connected";
+
+        static readonly string[] subscriptions = new[] { TopicDeviceConnected };
+
+        readonly IConnectionProvider connectionProvider;
+        readonly IIdentityProvider identityProvider;
+        readonly ISystemComponentIdProvider systemComponentIdProvider;
+        readonly DeviceProxy.Factory deviceProxyFactory;
+
+        readonly Channel<MqttPublishInfo> notifications;
+        readonly Task processingLoop;
+
+        AsyncLock guard = new AsyncLock();
+
+        // Normal dictionary would be sufficient because of the locks, however we need AddOrUpdate()
+        ConcurrentDictionary<IIdentity, IDeviceListener> knownConnections = new ConcurrentDictionary<IIdentity, IDeviceListener>();
+
+        // this class is auto-registered so no way to implement an async activator.
```

Comment:
What do you mean auto-registered? We have a pattern to do this with Autofac using a Create method..

Reply:
By auto-registered I mean that the ConnectionHandler is captured by its interfaces among other types (implementing the same interfaces), like the following (it is from MqttBridgeModule.cs, around line 38):

var componentTypes = MqttBridgeComponentDiscovery.GetCandidateTypes().ToArray();
builder.RegisterTypes(componentTypes)
                   .AsSelf()
                   .As(t => GetNonStandardBridgeInterfaces(t))
                   .SingleInstance();

Because of this, it cannot have that specific activator than many types in the xxxModule.cs files, with manually resolve their dependencies using 'await'.

One thing I can do is to change the 'auto-register' to a 'manual-register', meaning that I list all the classes manually, registering them - and in case of ConnectionHandler, I can create an async activator to avoid the x.Result in the code.

Just a side note, that using that many custom activators (basically manually resolve the dependencies of an instance) doesn't seem a good pattern - and it comes from a few Task<T> registrations. Later, every type using those Task<T>s needs a custom activator, and then the types using these later types need too, making a cascading need of custom activators. This decreases the power of dependency injection frameworks - just like in this current example. Maybe the long term solution would be to eliminate the Task<T> registrations.

Reply:
I agree with your last point, we need to look into how to make DI work with async creators. But passing in a Task in the ctor and then getting the Result doesn't seem like a good idea. So maybe we can just go with the current pattern for now..

Reply:
I fixed it the same than TwinController already does: it still gets a Task<T>, but instead of waiting in the constructor for the task to complete, it stores it as a "getter", and in the method it needs it (which is already an async menthod), it gets the connectionProvider with provider = await connectionProviderGetter (which is the Task<ConnectionProvider>)